### PR TITLE
8.0 add support for rental

### PR DESCRIPTION
**Pull request by alexis-via**
These rental modules were initially developed on https://code.launchpad.net/openerp-rental
So it's now time to move to github/OCA

### CODE REVIEW

File: sale_rental/__init__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from . import rental
+from . import wizard
```
Comment (bealdav) [inline sale_rental/__init__.py:24]: line 1 to 22 are not useful, just import no license mandatory, no explicit encoding, just ascii used
It's just a suggestion, you may keep as it if you want

File: sale_rental/README.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+=============
+
+In the menu *Sales > Products > Product Variants*, on the form view
+of a stockable product or consumable, in the *Rental* tab, there is a
+button *Create Rental Service* which starts a wizard to generate the
+corresponding rental service.
+
+In the menu *Warehouse > Configuration > Warehouses*, on the form view
+of the warehouse, in the *Technical Information* tab, you will see two
+additionnal stock locations: *Rental In* (stock of products to rent) and
```
Comment (JordiBForgeFlow) [inline sale_rental/README.rst]: Replace "additionnal" with "additional".

File: .travis.yml [hunk 3]
```diff
@@ -27,6 +27,8 @@ virtualenv:
 
 install:
   - git clone https://github.com/OCA/maintainer-quality-tools.git ${HOME}/maintainer-quality-tools
+  - git clone https://github.com/OCA/account-closing.git ${HOME}/account-closing
+  - git clone https://github.com/OCA/server-tools.git ${HOME}/server-tools
```
Comment (max3903) [inline .travis.yml]: Those repositories should be added to oca_dependencies.txt and don't need to be added here, I think.

File: sale_rental/i18n/fr.po [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+#. module: sale_rental
+#: selection:sale.order.line,rental_type:0
+msgid "New Rental"
+msgstr "Nouvelle location"
+
+#. module: sale_rental
+#: help:sale.rental,end_date:0
+msgid "End Date of the Rental, taking into account all the extensions sold to the customer."
+msgstr "End Date of the Rental, taking into account all the extensions sold to the customer."
```
Comment (bguillot) [inline sale_rental/i18n/fr.po]: Missing translation

File: sale_rental/i18n/fr.po [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+#. module: sale_rental
+#: field:sale.rental,out_state:0
+msgid "State of the Outgoing Stock Move"
+msgstr "État du mouvement de stock sortant"
+
+#. module: sale_rental
+#: code:addons/sale_rental/rental.py:406
+#, python-format
+msgid "The Rental Service of the Rental Extension you just selected is '%s' and it's not the same as the Product currently selected in this Sale Order Line."
+msgstr "The Rental Service of the Rental Extension you just selected is '%s' and it's not the same as the Product currently selected in this Sale Order Line."
```
Comment (bguillot) [inline sale_rental/i18n/fr.po]: Missing translation

File: sale_rental/i18n/fr.po [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+#. module: sale_rental
+#: selection:sale.rental,in_state:0
+#: selection:sale.rental,out_state:0
+msgid "Available"
+msgstr "Available"
+
+#. module: sale_rental
+#: model:stock.location,name:sale_rental.rental_out_stock_location
+msgid "Rental Out"
+msgstr "Rental Out"
```
Comment (bguillot) [inline sale_rental/i18n/fr.po]: Missing translations

File: sale_rental/i18n/fr.po [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+#. module: sale_rental
+#: view:sale.rental:0
+msgid "Return"
+msgstr "Retour"
+
+#. module: sale_rental
+#: code:addons/sale_rental/rental.py:59
+#, python-format
+msgid "The rental product '%s' must have the option ''Must Have Start and End Dates' checked."
+msgstr "The rental product '%s' must have the option ''Must Have Start and End Dates' checked."
```
Comment (bguillot) [inline sale_rental/i18n/fr.po]: Missing translation

**Comment by bguillot** [comment]
Thanks @alexis-via 

:+1:

File: sale_rental/wizard/create_rental_product.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            self.env.context['active_id'])
+        return _('Rental of a %s') % hw_product.name
+
+    @api.model
+    def _default_code(self):
+        assert self.env.context.get('active_model') == 'product.product',\
+            'Wrong underlying model, should be product.product'
+        hw_product = self.env['product.product'].browse(
+            self.env.context['active_id'])
+        return _('RENT-%s') % hw_product.default_code
```
Comment (JordiBForgeFlow) [inline sale_rental/wizard/create_rental_product.py]: The stockable product that is used as a reference to create the service rental could have the default code set as empty. In that case the current code sets "RENT-false". I suggest to add the default code for the created rental service only if the associated product has a default_code set.

Comment (JordiBForgeFlow) [inline sale_rental/wizard/create_rental_product.py]: I can see that the default code is a required field in the wizard. Why is that, if the default_code field is not a required field at all in the product? What if the original product has no default code?

File: sale_rental/rental.py [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+    can_sell_rental = fields.Boolean(string='Can Sell from Rental')
+    rental_type = fields.Selection([
+        ('new_rental', 'New Rental'),
+        ('rental_extension', 'Rental Extension'),
+        ], 'Rental Type',
+        readonly=True, states={'draft': [('readonly', False)]})
+    extension_rental_id = fields.Many2one(
+        'sale.rental', string='Rental to Extend')
+    rental_qty = fields.Float(
+        string='Rental Quantity', digits=dp.get_precision('Product UoS'))
```
Comment (JordiBForgeFlow) [inline sale_rental/rental.py]: I find confusing to see the following fields in the sale order line:
Number of Days
Rental Quantity
Quantity

I suggest that adding a tooltip to the field 'Rental Quantity' will help the user to clarify what to enter here.
For example:
help = "Indicate the number of items that will be delivered for rental."

File: sale_rental/wizard/create_rental_product.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    def create_rental_product(self):
+        self.ensure_one()
+        #  check that a rental product doesn't already exists ?
+        assert self.env.context.get('active_model') == 'product.product',\
+            'Wrong underlying model, should be product.product'
+        hw_product_id = self.env.context.get('active_id')
+        assert hw_product_id, 'Active ID is not set'
+        pp_obj = self.env['product.product']
+        day_uom_id = self.env.ref('product.product_uom_day').id
+
```
Comment (JordiBForgeFlow) [inline sale_rental/wizard/create_rental_product.py]: I suggest copying the image from the hw_product to the new rental service.

Add:
hw_product = self.env['product.product'].browse(
             hw_product_id)

File: sale_rental/wizard/create_rental_product.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            'sale_ok': True,
+            'purchase_ok': False,
+            'uom_id': day_uom_id,
+            'uom_po_id': day_uom_id,
+            'list_price': self.sale_price_per_day,
+            'name': self.name,
+            'default_code': self.default_code,
+            'rented_product_id': hw_product_id,
+            'must_have_dates': True,
+            'categ_id': self.categ_id.id,
```
Comment (JordiBForgeFlow) [inline sale_rental/wizard/create_rental_product.py:82]: To copy the image, add:
'image': hw_product.image,

**Comment by alexis-via** [comment]
@jbeficent I took into account your remarks. It deserves a :+1:  :)

**Comment by alexis-via** [comment]
@jbeficent Any other remarks ? We need a second positive review to merge this.

**Comment by JordiBForgeFlow** [comment]
I have some work in progress on tests and spanish translations. Will submit a separate PR.

Another remark, I tested with module 'stock_quant_manual_assign' from @pedrobaeza and it allows you to manually select the lot/serial that you want to rent. This is really a value added function, in cases where you rent serialized equipment.

**Comment by JordiBForgeFlow** [comment]
+1 to merge. Good work @alexis-via

**Comment by sebastienbeau** [comment]
good work :+1:

**Pull request merged**